### PR TITLE
Pipeline/llm

### PR DESCRIPTION
**Pull request by deep-diver**
This PR was tested with both 
- TGI(on Hugging Face Inference w/ PRO account)
- locally running Hugging Face model(w/ `transformers` on M3 Macbook Pro Max)

Some changes are for introducing Async feature.
The next step is to build Gradio app.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,31 +1,73 @@
...
+                "mistralai/Mixtral-8x7B-Instruct-v0.1",
+                "NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO",
+                "mistralai/Mistral-7B-Instruct-v0.2",
+                "mistralai/Mistral-7B-Instruct-v0.1",
+                "HuggingFaceH4/zephyr-7b-beta",
+                "meta-llama/Llama-2-7b-chat-hf",
+                "meta-llama/Llama-2-13b-chat-hf",
+                "meta-llama/Llama-2-70b-chat-hf",
+                "openchat/openchat-3.5-0106"
+            ]
```
Comment (sayakpaul) [inline main.py]: Should be treated as a constant.

File: main.py [hunk 2]
```diff
@@ -1,31 +1,73 @@
...
+            if args.model_id not in allowed_models_hf_pro:
+                raise ValueError("not supported model for Hugging Face PRO account")
         return func(args)
     return inner_function
 
 @validate_args
-def workflow(args):
```
Comment (sayakpaul) [inline main.py:13]: Seems like this function is for convenience as it doesn't return anything.

File: main.py [hunk 3]
```diff
@@ -1,31 +1,73 @@
...
     parser.add_argument('--prompt-tpl-path', type=str, default="vid2persona/prompts")
+    parser.add_argument('--hf-access-token', type=str, default=None)
+    parser.add_argument('--model-id', type=str, default="HuggingFaceH4/zephyr-7b-beta")
+    parser.add_argument('--max-input-token-length', type=int, default=4096)
+    parser.add_argument('--max-new-tokens', type=int, default=1024)
+    parser.add_argument('--temperature', type=float, default=0.6)
+    parser.add_argument('--top-p', type=float, default=0.9)
+    parser.add_argument('--top-k', type=int, default=50)
+    parser.add_argument('--repetition-penalty', type=float, default=1.2)
+    parser.add_argument('--message', type=str, default="Hello there! How are you doing?")
```
Comment (sayakpaul) [inline main.py]: Let's add `help` for these two. Rest I think are fairly self-explanatory.

File: vid2persona/gen/tgi_openllm.py [hunk 4]
```diff
@@ -1,114 +1,23 @@
...
-  }
-  data = {
-      'inputs': prompt,
-      'stream': True,
-      'options': {
-          'use_cache': False,
-      },
-      'parameters': parameters
-  }
+    client = AsyncOpenAI(
```
Comment (sayakpaul) [inline vid2persona/gen/tgi_openllm.py:13]: Didn't know about this one :o

**Review by sayakpaul — APPROVED** [review]
Superb work!

File: main.py (hunk 2 above)
Comment (deep-diver) [reply main.py:13]: yes. the `main.py` CLI is to verify if other parts are working fine as a library.

**Comment by deep-diver** [comment]
@sayakpaul 

I found it is difficult to limit the input tokens < `max-input-token-length`:

- For both TGI and locally running models, we can count the number of input tokens and trim them down < `max-input-token-length` just like how you did from the previous notebook.

- However, it loses some important information. For instance, `<|system|>`, `<|user|>`, and `<|assistant|>` special tokens gives the model signals. With naively trimming down `[-max-input-token-length:]`, we lose `<|system|>` part. 

- Even if we keep `<|system|>` part and trim the rest, we don't want to lose `<|user|>` and `<|assistant|>` special tokens as well. Meaning, the ideal final output should be something like below:

```
<|system|>
ALWAYS KEEP THIS PART
<|user|>
.....MIGHT NEED TO BE TRIMMED WITHIN <|user|> section, but we should keep <|user|> special token.
<|assistant|>
.....
<|user|>
WHATEVER USER's say
```

The main obstacle is to keep special tokens, and trim down the actual contents inside each special tokens (possible just remove the whole thing if the len(content) becomes zero after trimming. 

- Furthermore, even if we have successfully manage this trimming process, the problem is how we can convert the trimmed prompts back to the message format such as `[{"role": ..., "content": ...}, ...]`. This is necessary since OpenAI SDK only allows us to input the messages in such format. 
    - this issue introduces another issue that tokenized text now is a sort of raw text which includes different special tokens from model to model. Like, for `HuggingFaceH4/zephyr-7b-beta`, it has `<|system|>`, `<|user|>`, and `<|assistant|>` special tokens. However, for `NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO`, it has only `<|im_start|>` special token. etc.,


What do you think?
I mean we don't need to tackle this issue within this project, but wanted to discuss about what I have found so far. If you already know the solution for this, please let me know!

**Comment by sayakpaul** [comment]
Do you want to post it on the fellows channel and tag me?

**Pull request merged**